### PR TITLE
Table-Driven Integration Tests

### PR DESCRIPTION
**Pull request by d-ylee**
@vkuznet As I was moving towards doing the `datasets` endpoint for #4 , I decided to try the table-driven integration tests again. I simplified the test workflow to either GET or POST to the endpoint, depending on the configuration in the test case. Please let me know your thoughts.

### CODE REVIEW

File: test/integration_cases.go [hunk 1]
```diff
@@ -0,0 +1,782 @@
+package main
```
[the remark this answers is not in the document]
Comment (d-ylee) [reply test/integration_cases.go:1]: @vkuznet Thanks for the feedback. I did remember that in the initial issue #4 , you mentioned:

> generate CMS DBS meta-data using lexicon rules, such that we'll have random runs/files/blocks/datasets and inject them into server (current DBSClient provides that , see [here](https://github.com/dmwm/DBSClient/blob/main/tests/dbsclient_t/unittests/DBSClientWriter_t.py#L18)). We need to provide generator of JSON records which eventually we can use DBS POST APIs.

Would you suggest for us to mirror the same functionality as shown in DBSClientWriter_t.py (https://github.com/dmwm/DBSClient/blob/43016fd52529f864596f2ededaee04dcd75f20a4/tests/dbsclient_t/unittests/DBSClientWriter_t.py#L18-L49), or would be okay to use the same JSON data each time for the requests?

**Comment by vkuznet** [comment]
No, we should not do this as in Python DBSClientWriter_t.py because this defeat purpose of unit test. In order to test that API is working you must know the input, i.e. in other words you know what you injected, and you can use this information to tests that you APIs will deliver you the same. Therefore, I suggest to have static JSON files, which you can use for both injection phase when you'll use DBS Writer to inject the data, and then you can use these files for reading phase, i.e. when you use DBSReader to look-up the data. Since we can do all testing in SQLiteDB and don't really need ORACLE we don't need to regenerate data over and over again (this is the reason DBSClientWriter_t.py was written).

**Comment by d-ylee** [comment]
@vkuznet I updated the test with the recommended changes.

I do have a question about error responses. When I do a GET request with `/primarydstypes?datasets=fnal`, the response is an expected error due to "fnal" not being a valid pattern/query text. I did notice that the code in the `error.reason` and the `error.code` and `message` in the response are different (113 vs 114). Is this expected?

```javascript
[{
    "error": {
        "reason": "DBSError Code:114 Description:DBS validation error when wrong pattern is provided Function:dbs.validator.Check Message:unable to match 'dataset' value 'fnal' Error: invalid parameter(s)",
        "message": "not str type",
        "function": "dbs.Validate",
        "code": 113
    },
    http: "...",
    "exception": 400,
    "type": "HTTPError",
    "message": "DBSError Code:113 Description:DBS validation error, e.g. input parameter does not match lexicon rules Function:dbs.Validate Message:not str type Error: nested DBSError Code:114 Description:DBS validation error when wrong pattern is provided Function:dbs.validator.Check Message:unable to match 'dataset' value 'fnal' Error: invalid parameter(s)"
}]
```

**Comment by d-ylee** [comment]
@vkuznet What are your thoughts on the changes?

File: test/integration_cases.go [hunk 2]
```diff
@@ -0,0 +1,1006 @@
...
+
+// basic elements to define a test case
+type testCase struct {
+	description string     // test case description
+	serverType  string     //DBSWriter, DBSReader, DBSMigrate
+	method      string     // http method
+	endpoint    string     // url endpoint
+	params      url.Values // url parameters
+	handler     func(http.ResponseWriter, *http.Request)
+	input       RequestBody //kjdPOST record
```
Comment (vkuznet) [inline test/integration_cases.go:40]: Please fix the comment `//kjdPOST record`

**Review by vkuznet — CHANGES_REQUESTED** [review]
I only pointed to few minor issues which you should fix. But I also suggest to split `test/integration_cases.go` file which is becoming too large into individual files which would be associated with specific DBS table. For instance, you may put getPhysicsGroupsTestTable function into `test/int_physics_groups_case.go` file. Since we have plenty of tables to test we'll need code for them and keeping all of these functions in a single file become tedious. Just use common convention, similar to what you see in `dbs` area.

**Review by vkuznet — CHANGES_REQUESTED** [review]
Dennis, I think everything is fine, but I still want to suggest few things:
- let's reduce file names, instead of `int_<>_cases.go` we should have `int_<>.go`, e.g. rename `int_processing_eras_cases.go` to `int_processing_eras.go`. Think `_cases` is redundant in file names here.
- I suggest to add proper doc string in each `int_<>.go` file with reference to main `integration_cases.go` file. This will help to navigate the code. For instance we may add the following comment (e.g. processing_eras):
```
// this file contains integration logic for processing_eras
// the HTTP requests is defined by dbs.processingEras struct defined in dbs/processingeras.go file
// the HTTP response is defined by dbs.ProcessingEra struct defined in dbs/...go file
// the HTTP endpoint is defined by EndpointTestCase struct defined in test/integration_cases.go file
```
- I also suggest to add comprehensive documentation to test/integration_cases.go to explain the entire workflow. Please put all details how test should be setup and run, keep in mind that it should be clear to non-expert or to a person who will see it first time. This is important to allow others to contribute to the codebase and run the tests.

After that I think we can merge your PR

File: test/int_acquisition_eras.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
+package main
+
+// this file contains logic for acquisitioneras API
+// the HTTP requests body is defined by dbs.AcquisitionEras struct defined in dbs/acquisition_eras.go
```
Comment (vkuznet) [inline test/int_acquisition_eras.go:4]: the file is dbs/acquisitioneras.go

File: test/int_primary_datasets.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
+package main
+
+// this file contains logic for primarydatasets and primarydstypes API
+//
+// primarydatasets:
+// the HTTP request body is defined by dbs.PrimaryDatasetRecord struct defined in dbs/primarydatasets.go
+// the HTTP response body is defined by dbs.PrimaryDataset struct defined in this dbs/primarydatasets.go
```
Comment (vkuznet) [inline test/int_primary_datasets.go:7]: there is no `dbs.PrimaryDataset`, where does it come from?

File: test/int_primary_datasets.go [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+// primarydataset and primarydstype endpoints tests
+func getPrimaryDatasetTestTable(t *testing.T) EndpointTestCase {
+	// create data structs for expected requests and responses
+	primaryDSReq := dbs.PrimaryDatasetRecord{
+		PRIMARY_DS_NAME: TestData.PrimaryDSName,
+		PRIMARY_DS_TYPE: "test",
+		CREATE_BY:       "tester",
+	}
+	primaryDSResp := dbs.PrimaryDataset{
```
Comment (vkuznet) [inline test/int_primary_datasets.go:37]: there is no `dbs.PrimaryDataset` in current codebase, may be you added it in your fork and forget to add to this commit.

File: test/int_processing_eras.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
+package main
+
+// this file contains logic for processingeras API
+// the HTTP request body is defined by dbs.ProcessingEras struct defined in dbs/processingeras.go
+// the HTTP response body is defined by dbs.ProcessingEra struct defined in dbs/processingeras.go
```
Comment (vkuznet) [inline test/int_processing_eras.go:5]: there is no `dbs.ProcessingEra` in current codebase, may be you added it in your fork and forget to add to this commit.

File: test/int_processing_eras.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+// processingeras endpoint tests
+// TODO: Rest of test cases
+func getProcessingErasTestTable(t *testing.T) EndpointTestCase {
+	procErasReq := dbs.ProcessingEras{
+		PROCESSING_VERSION: int64(TestData.ProcessingVersion),
+		DESCRIPTION:        "this_is_a_test",
+		CREATE_BY:          "tester",
+	}
+	procErasResp := dbs.ProcessingEra{
```
Comment (vkuznet) [inline test/int_processing_eras.go:24]: there is no `dbs.ProcessingEra` in current codebase, may be you added it in your fork and forget to add to this commit.

**Comment by vkuznet** [comment]
that's fine but you should clear up comments about it. And, you need to understand the difference. For instance, `dbs.AcquisitionEras` represents data structure which reflects the DBS table, while `dbs.AcquisitionEra` is what `bulkblocks` API use to provide to end-user. In this case the difference is `acquisition_era_id` which is not part of `bulkblocks` API output, but it needs to be used in queries, joins etc. I have a plan to separate API implementation from data structs and put the later into separate files, but this will be done later once we get full functionality in place.

**Comment by vkuznet** [comment]
please update docs accordingly that we'll have everything in order.

**Comment by vkuznet** [comment]
ok, everything is fine, I'll merge it. THANKS!!!

**Pull request merged**